### PR TITLE
processors/baggage: add baggage span processor

### LINKED ISSUE #5397: Request for new component: Baggage Span Processor
The Honeycomb distro currently provides a Span processor that takes items from the baggage and adds those items as attributes onto spans. We'd like to donate this code upstream if there's an interest from the community, the code is pretty minimal: https://github.com/honeycombio/honeycomb-opentelemetry-go/blob/main/baggage_span_processor.go

### PR DESCRIPTION
**Pull request by codeboten**
Fixes #5397

### CODE REVIEW

File: processors/baggage/processor_test.go
Comment (pellared) [inline processors/baggage/processor_test.go:40]: Use https://pkg.go.dev/go.opentelemetry.io/otel/baggage#NewMemberRaw instead.

Comment (pellared) [inline processors/baggage/processor_test.go:40]: Can you `require.NoError`?

File: processors/baggage/processor_test.go
Comment (pellared) [inline processors/baggage/processor_test.go:41]: Can you `require.NoError`?

File: processors/baggage/processor_test.go
Comment (pellared) [inline processors/baggage/processor_test.go:39]: Can we use https://pkg.go.dev/go.opentelemetry.io/otel/baggage#New instead?

File: processors/baggage/processor_test.go
Comment (pellared) [inline processors/baggage/processor_test.go:57]: You can use `assert.Len` instead.

File: processors/baggage/processor_test.go
Comment (pellared) [inline processors/baggage/processor_test.go:62]: AFAIK you can simply

```suggestion
	want := []attribute.KeyValue{ attribute.String("baggage.test", "baggage value") }
	assert.Equal(t, want, exporter.spans[0].Attributes())
```

File: processors/baggage/processor.go
Comment (pellared) [inline processors/baggage/processor.go:14]: Can we export the type so that we can possibly add additional methods later if needed?

File: processors/baggage/processor_test.go
Comment (codeboten) [reply processors/baggage/processor_test.go]: Done 👍🏻

File: processors/baggage/processor_test.go
Comment (codeboten) [reply processors/baggage/processor_test.go]: Done 👍🏻

File: processors/baggage/processor_test.go
Comment (codeboten) [reply processors/baggage/processor_test.go]: Done 👍🏻

File: processors/baggage/processor_test.go
Comment (codeboten) [reply processors/baggage/processor_test.go]: Done

File: processors/baggage/processor.go
Comment (codeboten) [reply processors/baggage/processor.go:14]: @pellared we could, though I'm not sure if there's a need just yet. I don't have a strong preference one way or the other, happy to follow your guidance on this.

Comment (pellared) [reply processors/baggage/processor.go:14]: > though I'm not sure if there's a need just yet

It is in order to avoid making a breaking change.

Lets rename it to


```suggestion
type SpanProcessor struct{}
```

Comment (codeboten) [reply processors/baggage/processor.go:14]: 👍🏻

Comment (codeboten) [reply processors/baggage/processor.go:14]: Renamed and made public.

**Comment by pellared** [comment]
Can you also update the `README.md`? I see this is the first processor. Shouldn't we have a "split" for `trace` and `log` processors?

**Comment by codeboten** [comment]
@pellared i've updated the readme and moved the contents of the `baggage` package for traces under `baggage/baggagetrace` to follow the exporters pattern in the go repo.

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:16]: We're usually linking the PR doing the change, not the related issue.

```suggestion
  This module provides a Baggage Span Processor. (#5404)
```

**Review by dmathieu — COMMENTED** [review]
How about adding an example_test?

File: processors/baggage/baggagetrace/doc.go
Comment (pellared) [inline processors/baggage/baggagetrace/doc.go:6]: This comment would better fit as a comment for the `SpanProcessor` type.

**Review by pellared — COMMENTED** [review]
Just nit comments to address

Overall, LGTM

File: CHANGELOG.md
Comment (codeboten) [inline CHANGELOG.md:16]:
```suggestion
- The `go.opentelemetry.io/contrib/processors/baggage` module. This module provides a Baggage Span Processor. (#5404)
```

File: processors/baggage/baggagetrace/doc.go
Comment (codeboten) [inline processors/baggage/baggagetrace/doc.go:6]:
```suggestion
// Package baggagetrace implements a baggage span processor.
```

File: processors/baggage/baggagetrace/go.mod
Comment (pellared) [inline processors/baggage/baggagetrace/go.mod:1]:
```suggestion
module go.opentelemetry.io/contrib/processors/baggage/baggagetrace
```

File: CHANGELOG.md
Comment (pellared) [inline CHANGELOG.md:15]:
```suggestion
- The `go.opentelemetry.io/contrib/processors/baggage/baggagetrace` module. This module provides a Baggage Span Processor. (#5404)
```

File: processors/baggage/baggagetrace/processor.go
Comment (pellared) [inline processors/baggage/baggagetrace/processor.go:4]:
```suggestion
package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage/baggagetrace"
```

File: processors/baggage/baggagetrace/doc.go
Comment (pellared) [inline processors/baggage/baggagetrace/doc.go:5]:
```suggestion
package baggagetrace // import "go.opentelemetry.io/contrib/processors/baggage/baggagetrace"
```

File: CODEOWNERS
Comment (codeboten) [inline CODEOWNERS:49]:
```suggestion
processors/baggage/baggagetrace                                         @open-telemetry/go-approvers @codeboten
```

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Comment by pellared** [comment]
> I see this is approved, was there more work needed before merging this?

It was discussed during last SIG meetings. We just want to have more reviews.

**Review by MikeGoldsmith — APPROVED** [review]

File: CODEOWNERS
Comment (MikeGoldsmith) [inline CODEOWNERS:49]: Happy to add my name to the component owners sections too.

```suggestion
processors/baggage/baggagetrace                                         @open-telemetry/go-approvers @codeboten @MikeGoldsmith
```

**Comment by MikeGoldsmith** [comment]
> > I see this is approved, was there more work needed before merging this?
> 
> It was discussed during last SIG meetings. We just want to have more reviews.

Hey @pellared - what is the best way to get more community reviews? This PR has been open for more than two weeks and only a couple of approvers have provided feedback.

Do you have a sense how many reviews and/or how long you'd want to wait for additional reviews before accepting?

Is it specific people whom you would like to provide feedback?

**Comment by MikeGoldsmith** [comment]
FYI the CI failure is from codecob being rate limited
```
debug - 2024-05-02 15:31:38,221 -- Commit creating result --- {"result": "RequestResult(error=RequestError(code='HTTP Error 429', params={}, description='{\"detail\":\"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds.\"}'), warnings=[], status_code=429, text='{\"detail\":\"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds.\"}')"}
error - 2024-05-02 15:31:38,221 -- Commit creating failed: {"detail":"Tokenless has reached GitHub rate limit. Please upload using a token: https://docs.codecov.com/docs/adding-the-codecov-token. Expected available in 491 seconds."}
```

**Review by MadVikingGod — APPROVED** [review]

**Comment by pellared** [comment]
@MrAlias, if I remember correctly you wanted to review this PR as well.

**Pull request merged**
